packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity/generate-property-row-recursively.ts

```diff
@@ -0,0 +1,103 @@
+import { isPlainObject } from "lodash";
+import {
+  getPropertyTypesByBaseUri,
+  RootEntityAndSubgraph,
+} from "../../../../../../../../../lib/subgraph";
+import { PropertyRow } from "../../types";
+import { getDataTypesOfPropertyType } from "./get-data-types-of-property-type";
+
+/**
+ * This function generates property row data,
+ * and calls itself again for each nested property. Then puts results of these recursive calls into `children` array
+ *
+ * @param properties
+ * Properties object for current depth. On entity level, it starts from `entity.properties`
+ *
+ * @param propertyTypeBaseUri
+ * Name of the specific property inside `properties` object
+ *
+ * @param propertyKeyChain
+ * An array of `propertyTypeBaseUri`'s to store the path to each nested property.
+ * ```js
+ * // Example: `propertyKeyChain` of `c`
+ * properties = { a: { b: { c: "John" } } };
+ * propertyKeyChain = ["a", "b", "c"]
+ * ```
+ *
+ * @param rootEntityAndSubgraph
+ * An object storing root entity & subgraph of that entity
+ *
+ * @param requiredPropertyTypes
+ * An array of `propertyTypeBaseUri`'s.
+ * This is used to check if a property should be shown as `required` or not
+ *
+ * @param depth
+ * Depth of the property. For properties at root, depth starts from `0`
+ *
+ * @returns property row (and nested rows as `children` if it's a nested property)
+ */
+export const generatePropertyRowRecursively = (
+  properties: any,
```

Comment:
Tried to change this, but got stuck with typescript. It felt OK to use `any` here for me, since it could by anything nested.

If it's unsafe to use any, and we should switch to `unknown`, I'd appreciate some help about this chaneg.

Reply:
This patch will update this to `any`, and also the change to removing the need for `propertyKeyChain` as mentioned elsewhere. 

The main trick is using `isUnknownObject`, which lets you cast `unknown` to an object of `unknown` properties. 

Let me know if you have any questions

```diff
Index: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-on-cell-edited.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-on-cell-edited.ts b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-on-cell-edited.ts
--- a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-on-cell-edited.ts	(revision 0d27534690cd698bfff0a1e13cb883f92c0d9afb)
+++ b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-on-cell-edited.ts	(date 1667410226660)
@@ -44,14 +44,13 @@
 
       const updatedProperties = cloneDeep(entity.properties);
 
-      const { propertyKeyChain } = property;
-
       /**
        * we're reaching to the nested property by the property keys array
-       * so we can update the deeply nested properties,
-       * by using keys pathing to specific property called `propertyKeyChain`
+       * so we can update the deeply nested properties – the rowId is equivalent
+       * to its dot-separated property path which lodash
+       * can use to update
        */
-      set(updatedProperties, propertyKeyChain, valueCell.data.property.value);
+      set(updatedProperties, property.rowId, valueCell.data.property.value);
 
       /**
        * setting state for optimistic update
Index: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity.ts b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity.ts
--- a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity.ts	(revision 0d27534690cd698bfff0a1e13cb883f92c0d9afb)
+++ b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity.ts	(date 1667409965481)
@@ -21,10 +21,8 @@
     generatePropertyRowRecursively(
       properties,
       propertyTypeBaseUri,
-      [propertyTypeBaseUri],
       rootEntityAndSubgraph,
       requiredPropertyTypes,
-      0,
     ),
   );
 };
Index: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity/generate-property-row-recursively.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity/generate-property-row-recursively.ts b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity/generate-property-row-recursively.ts
--- a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity/generate-property-row-recursively.ts	(revision 0d27534690cd698bfff0a1e13cb883f92c0d9afb)
+++ b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity/generate-property-row-recursively.ts	(date 1667410349792)
@@ -1,3 +1,4 @@
+import { isUnknownObject } from "@hashintel/hash-shared/util";
 import { isPlainObject } from "lodash";
 import {
   getPropertyTypesByBaseUri,
@@ -16,14 +17,6 @@
  * @param propertyTypeBaseUri
  * Name of the specific property inside `properties` object
  *
- * @param propertyKeyChain
- * An array of `propertyTypeBaseUri`'s to store the path to each nested property.
- * ```js
- * // Example: `propertyKeyChain` of `c`
- * properties = { a: { b: { c: "John" } } };
- * propertyKeyChain = ["a", "b", "c"]
- * ```
- *
  * @param rootEntityAndSubgraph
  * An object storing root entity & subgraph of that entity
  *
@@ -31,17 +24,21 @@
  * An array of `propertyTypeBaseUri`'s.
  * This is used to check if a property should be shown as `required` or not
  *
+ * @param parentPath
+ * The dot-separated path of the parent property. For properties at root,
+ * parentPath starts with `null`.
+ *
  * @param depth
  * Depth of the property. For properties at root, depth starts from `0`
  *
  * @returns property row (and nested rows as `children` if it's a nested property)
  */
 export const generatePropertyRowRecursively = (
-  properties: any,
+  properties: unknown,
   propertyTypeBaseUri: string,
-  propertyKeyChain: string[],
   rootEntityAndSubgraph: RootEntityAndSubgraph,
   requiredPropertyTypes: string[],
+  parentPath: string | null = null,
   depth = 0,
 ): PropertyRow => {
   const propertyTypeVersions = getPropertyTypesByBaseUri(
@@ -64,31 +61,36 @@
 
   const required = !!requiredPropertyTypes?.includes(propertyTypeBaseUri);
 
-  const value = properties[propertyTypeBaseUri];
+  let value;
 
   const children: PropertyRow[] = [];
 
-  // generate rows for nested properties and push them to children array
-  if (isPlainObject(value)) {
-    for (const subPropertyTypeBaseUri of Object.keys(value)) {
-      children.push(
-        generatePropertyRowRecursively(
-          properties[propertyTypeBaseUri],
-          subPropertyTypeBaseUri,
-          [...propertyKeyChain, subPropertyTypeBaseUri],
-          rootEntityAndSubgraph,
-          requiredPropertyTypes,
-        ),
-      );
+  const ourPath = `${parentPath ? `${parentPath}.` : ""}${propertyTypeBaseUri}`;
+
+  if (isUnknownObject(properties)) {
+    value = properties[propertyTypeBaseUri];
+
+    // generate rows for nested properties and push them to children array
+    if (isPlainObject(value) && isUnknownObject(value)) {
+      for (const subPropertyTypeBaseUri of Object.keys(value)) {
+        children.push(
+          generatePropertyRowRecursively(
+            properties[propertyTypeBaseUri],
+            subPropertyTypeBaseUri,
+            rootEntityAndSubgraph,
+            requiredPropertyTypes,
+            ourPath,
+            depth + 1,
+          ),
+        );
+      }
     }
   }
 
   const indent = !depth ? 0 : children.length ? depth : depth - 1;
 
-  const rowId = propertyKeyChain.join(".");
-
   return {
-    rowId,
+    rowId: ourPath,
     title: propertyType.title,
     value,
     dataTypes,
@@ -102,6 +104,5 @@
      * and check some values of prev-next items on the flattened tree while calculating this
      */
     verticalLinesForEachIndent: [],
-    propertyKeyChain,
   };
 };
Index: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/types.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/types.ts b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/types.ts
--- a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/types.ts	(revision 0d27534690cd698bfff0a1e13cb883f92c0d9afb)
+++ b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/types.ts	(date 1667410349554)
@@ -3,6 +3,7 @@
 
 export type PropertyRow = {
   title: string;
+  // This must be equivalent to the dot-separated path to the property
   rowId: string;
   value: unknown;
   dataTypes: string[];
@@ -11,7 +12,6 @@
   depth: number;
   indent: number;
   verticalLinesForEachIndent: VerticalIndentationLineDir[];
-  propertyKeyChain: string[];
 };
 
 export type PropertyColumnKey = Extract<
```

Reply:
Updated to `unknown` here https://github.com/hashintel/hash/pull/1266/commits/ee37ef64e5f928bcf58a82bd66469b64396ac529

Check this comment for regarding `propertyKeyChain` https://github.com/hashintel/hash/pull/1266#discussion_r1012484305